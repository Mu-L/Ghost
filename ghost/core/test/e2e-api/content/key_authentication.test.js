const {agentProvider, fixtureManager, matchers, configUtils} = require('../../utils/e2e-framework');
const {anyContentVersion, anyEtag, anyErrorId} = matchers;

describe('Content API key authentication', function () {
    describe('Unauthenticated', function () {
        let unauthenticatedAgent;

        before(async function () {
            unauthenticatedAgent = await agentProvider.getContentAPIAgent();
            await fixtureManager.init('api_keys');
        });

        it('Can not access without key', async function () {
            await unauthenticatedAgent
                .get('posts/')
                .expectStatus(403)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId
                    }]
                });
        });
    });

    describe('Authenticated', function () {
        let authenticatedAgent;

        before(async function () {
            authenticatedAgent = await agentProvider.getContentAPIAgent();
            await fixtureManager.init('api_keys');
            authenticatedAgent.authenticate();
        });

        it('Can access with with valid key', async function () {
            await authenticatedAgent
                .get('posts/')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                });
        });
    });

    describe('Host Settings: custom integration limits', function () {
        let limitedAgent;

        before(async function () {
            await configUtils.set('hostSettings:limits', {
                customIntegrations: {
                    disabled: true,
                    error: 'Custom limit error message'
                }
            });

            limitedAgent = await agentProvider.getContentAPIAgent();
            await fixtureManager.init('integrations', 'api_keys');
            limitedAgent.authenticate();
        });

        afterEach(async function () {
            await configUtils.restore();
        });

        it('Blocks the request when host limit is in place for custom integrations', async function () {
            await limitedAgent
                .get('posts/')
                .expectStatus(403)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId,
                        type: 'HostLimitError',
                        message: 'Custom limit error message'
                    }]
                });

            // CASE: explore endpoint can only be reached by Admin API
            await limitedAgent
                .get('explore/')
                .expectStatus(404)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    errors: [{
                        id: anyErrorId,
                        type: 'NotFoundError'
                    }]
                });
        });
    });
});
