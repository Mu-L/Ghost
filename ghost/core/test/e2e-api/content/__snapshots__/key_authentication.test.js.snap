// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content API key authentication Authenticated Can access with with valid key 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "36545",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Content API key authentication Host Settings: custom integration limits Blocks the request when host limit is in place for custom integrations 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": Object {
        "name": "customIntegrations",
      },
      "ghostErrorCode": null,
      "help": "https://ghost.org/help/",
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Custom limit error message",
      "property": null,
      "type": "HostLimitError",
    },
  ],
}
`;

exports[`Content API key authentication Host Settings: custom integration limits Blocks the request when host limit is in place for custom integrations 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "259",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Content API key authentication Host Settings: custom integration limits Blocks the request when host limit is in place for custom integrations 3: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Resource not found",
      "property": null,
      "type": "NotFoundError",
    },
  ],
}
`;

exports[`Content API key authentication Host Settings: custom integration limits Blocks the request when host limit is in place for custom integrations 4: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, max-age=0, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "204",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Content API key authentication Unauthenticated Can not access without key 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Unable to determine the authenticated member or integration. Check the supplied Content API Key and ensure cookies are being passed through if member auth is failing.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Authorization failed",
      "property": null,
      "type": "NoPermissionError",
    },
  ],
}
`;

exports[`Content API key authentication Unauthenticated Can not access without key 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "374",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
